File: packages/python/acs_cli/utils/detect.py
"""Framework and language detection from project files."""

import json
import os
import re
from pathlib import Path


JS_FRAMEWORK_DEPS = {
    "next": "nextjs",
    "react": "react",
    "vue": "vue",
    "nuxt": "nuxt",
    "svelte": "svelte",
    "@sveltejs/kit": "sveltekit",
    "angular": "angular",
    "@angular/core": "angular",
    "express": "express",
    "fastify": "fastify",
    "hono": "hono",
    "remix": "remix",
    "astro": "astro",
}

PYTHON_FRAMEWORKS = ["django", "fastapi", "flask", "tornado", "aiohttp"]


def slugify(name: str) -> str:
    slug = re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")
    return slug[:64]


def detect_project(directory: str = None) -> dict:
    """Return {'name': str, 'language': str|None, 'framework': str|None}."""
    if directory is None:
        directory = os.getcwd()

    name = slugify(Path(directory).name)

    # Node.js / TypeScript
    pkg_path = Path(directory) / "package.json"
    if pkg_path.exists():
        try:
            pkg = json.loads(pkg_path.read_text(encoding="utf-8"))
            all_deps = {**pkg.get("dependencies", {}), **pkg.get("devDependencies", {})}

            framework = None
            for dep, fw in JS_FRAMEWORK_DEPS.items():
                if dep in all_deps:
                    framework = fw
                    break

            has_ts = (
                "typescript" in all_deps
                or "ts-node" in all_deps
                or (Path(directory) / "tsconfig.json").exists()
            )

            pkg_name = slugify(pkg.get("name", name)) or name
            return {
                "name": pkg_name,
                "language": "typescript" if has_ts else "javascript",
                "framework": framework,
            }
        except (json.JSONDecodeError, OSError):
            pass

    # Python
    pyproject = Path(directory) / "pyproject.toml"
    requirements = Path(directory) / "requirements.txt"
    if pyproject.exists() or requirements.exists():
        framework = None
        if requirements.exists():
            reqs = requirements.read_text(encoding="utf-8").lower()
            for fw in PYTHON_FRAMEWORKS:
                if fw in reqs:
                    framework = fw
                    break
        return {"name": name, "language": "python", "framework": framework}

    # Rust
    if (Path(directory) / "Cargo.toml").exists():
        return {"name": name, "language": "rust", "framework": None}

    # Go
    if (Path(directory) / "go.mod").exists():
        return {"name": name, "language": "go", "framework": None}

    return {"name": name, "language": None, "framework": None}

File: cli/src/utils/detect.ts
import * as fs from "fs";
import * as path from "path";

export interface DetectedProject {
  name: string;
  language: string | null;
  framework: string | null;
}

const JS_FRAMEWORK_DEPS: Record<string, string> = {
  next: "nextjs",
  "next.js": "nextjs",
  react: "react",
  vue: "vue",
  nuxt: "nuxt",
  svelte: "svelte",
  "@sveltejs/kit": "sveltekit",
  angular: "angular",
  "@angular/core": "angular",
  express: "express",
  fastify: "fastify",
  hono: "hono",
  remix: "remix",
  "@remix-run/node": "remix",
  astro: "astro",
};

function slugify(str: string): string {
  return str
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, 64);
}

export function detectProject(dir: string = process.cwd()): DetectedProject {
  const name = slugify(path.basename(dir));

  // Node.js / TypeScript
  const pkgPath = path.join(dir, "package.json");
  if (fs.existsSync(pkgPath)) {
    try {
      const pkg = JSON.parse(fs.readFileSync(pkgPath, "utf8"));
      const allDeps = {
        ...pkg.dependencies,
        ...pkg.devDependencies,
      };

      let framework: string | null = null;
      for (const [dep, fw] of Object.entries(JS_FRAMEWORK_DEPS)) {
        if (allDeps[dep]) {
          framework = fw;
          break;
        }
      }

      const hasTypeScript =
        allDeps["typescript"] ||
        allDeps["ts-node"] ||
        fs.existsSync(path.join(dir, "tsconfig.json"));

      return {
        name: slugify(pkg.name || name),
        language: hasTypeScript ? "typescript" : "javascript",
        framework,
      };
    } catch {
      // fall through
    }
  }

  // Python
  const pyprojectPath = path.join(dir, "pyproject.toml");
  const requirementsPath = path.join(dir, "requirements.txt");
  if (fs.existsSync(pyprojectPath) || fs.existsSync(requirementsPath)) {
    let framework: string | null = null;
    if (fs.existsSync(requirementsPath)) {
      const reqs = fs.readFileSync(requirementsPath, "utf8").toLowerCase();
      if (reqs.includes("django")) framework = "django";
      else if (reqs.includes("fastapi")) framework = "fastapi";
      else if (reqs.includes("flask")) framework = "flask";
    }
    return { name, language: "python", framework };
  }

  // Rust
  const cargoPath = path.join(dir, "Cargo.toml");
  if (fs.existsSync(cargoPath)) {
    return { name, language: "rust", framework: null };
  }

  // Go
  const goModPath = path.join(dir, "go.mod");
  if (fs.existsSync(goModPath)) {
    return { name, language: "go", framework: null };
  }

  return { name, language: null, framework: null };
}
